File: JavaScript/7js 1/drc.js
/*
|--------------------------------------------------------------------------
| Resistor Color Codes
|--------------------------------------------------------------------------
|
*/

function decodeResistorColors(bands) {
   let color = ["black", "brown", "red", "orange", "yellow", "green", "blue", "violet", "gray", "white"];
   let band = bands.split(" ");
   let ohms1 =
      Number(String(color.indexOf(band[0])) + String(color.indexOf(band[1]))) *
      10 ** color.indexOf(band[2]);
   let ohms2 =
      ohms1 < 1000
         ? `${ohms1} ohms`
         : ohms1 >= 1000000
         ? `${ohms1 / 1000000}M ohms`
         : `${ohms1 / 1000}k ohms`;
   return band.length == 3 ? `${ohms2}, 20%` : band[3] == "gold" ? `${ohms2}, 5%` : `${ohms2}, 10%`;
}


function decodeResistorColors(bands) {
   const colors = {black: 0, brown: 1, red: 2, orange: 3, yellow: 4, green: 5, blue: 6, violet: 7, gray: 8, white: 9, gold: 5, silver: 10},
      [first, second, multiplier, tolerance] = bands.split(" ").map((color) => colors[color]),
      ohm = (10 * first + second) * 10 ** multiplier;
   return `${ohm >= 10 ** 6 ? ohm / 10 ** 6 + "M" : ohm >= 1000 ? ohm / 1000 + "k" : ohm} ohms, ${
      tolerance || 20
   }%`;
}


function decodeResistorColors(bands) {
   const obj = {black: 0, brown: 1, red: 2, orange: 3, yellow: 4, green: 5, blue: 6, violet: 7, gray: 8, white: 9, gold: 5, silver: 10};
   const arr = bands.split(` `);
   const res = (10 * obj[arr[0]] + obj[arr[1]]) * 10 ** obj[arr[2]];
   return `${res >= 1e6 ? `${res / 1e6}M` : res >= 1e3 ? `${res / 1e3}k` : res} ohms, ${
      obj[arr[3]] || 20
   }%`;
}


console.log(decodeResistorColors("yellow violet black"), "47 ohms, 20%");
console.log(decodeResistorColors("yellow violet red gold"), "4.7k ohms, 5%");
console.log(decodeResistorColors("brown black green silver"), "1M ohms, 10%");

File: Python/7py 1/drc.py
"""
|--------------------------------------------------------------------------
| Resistor Color Codes
|--------------------------------------------------------------------------
|
"""

def decode_resistor_colors(bands):
   color = {'black':0, 'brown':1, 'red':2, 'orange':3, 'yellow':4, 'green':5, 'blue':6, 'violet':7, 'gray':8, 'white':9, 'silver':10, 'gold':5}
   bands = [color[i] for i in bands.split()]
   ohms = (bands[0] * 10 + bands[1]) * 10 ** bands[2]
   ohms, x = (ohms/1000000.0, 'M') if ohms >= 1000000 else (ohms/1000.0, 'k') if ohms >= 1000 else (ohms, '')
   return "{}{} ohms, {}%".format(int(ohms) if ohms//1 == ohms else ohms, x, bands[3] if len(bands) > 3 else 20)


def decode_resistor_colors(bands):
   codes = {'black':0, 'brown':1, 'red':2, 'orange':3, 'yellow':4, 'green':5, 'blue':6, 'violet':7, 'gray':8, 'white':9, 'silver':10, 'gold':5}
   bands = [codes[band] for band in bands.split()] + [20]
   ohms = (bands[0] * 10 + bands[1]) * 10 ** bands[2]
   p = ''
   for c in 'kM':
      if ohms // 1000: ohms /= 1000; p = c
   return '%s%s ohms, %s%%' % (str(ohms).replace('.0', ''), p, bands[3])


def decode_resistor_colors(bands):
   color = {"black" : 0, "brown" : 1, "red" : 2, "orange" : 3, "yellow" : 4, "green" : 5, "blue" : 6, "violet" : 7, "gray" : 8, "white" : 9}
   tolerance = {"gold" : 5, "silver" : 10, "none" : 20}
   a, b, p, t, *_ = bands.split() + ["none"]
   c = (10 * color[a] + color[b]) * 10 ** color[p]
   r, m = next((c / x, y) for x, y in [(10 ** 6, "M"), (10 ** 3, "k"), (1, "")] if c // x > 0)
   return "{:g}{} ohms, {}%".format(r, m, tolerance[t])


print(decode_resistor_colors("yellow violet black"), "47 ohms, 20%")
print(decode_resistor_colors("yellow violet red gold"), "4.7k ohms, 5%")
print(decode_resistor_colors("brown black green silver"), "1M ohms, 10%")